[Playwright] DSM-PEPPER-777  kit upload send and receive a clinical saliva kit

GiCharkviani opened the pull request:
[PEPPER-777](https://broadworkbench.atlassian.net/jira/software/projects/pepper/boards/164?assignee=621677042d057500723f90e4&selectedIssue=PEPPER-777)

**All mandatory and possible assertions were made according to the ticket** 

[video.webm](https://github.com/broadinstitute/ddp-angular/assets/77500504/d7cae780-638d-4b15-8f0a-41c2ee88d8e3)


[PEPPER-777]: https://broadworkbench.atlassian.net/browse/PEPPER-777?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
DSM-Playwright-scanner-pages-and-more the first success

Commit:
DSM-Playwright-scanner-pages-and-more select-deselect changed

Commit:
DSM-Playwright-scanner-pages-and-more closer to the end

Commit:
DSM-Playwright-scanner-pages-and-more linter fix

Commit:
DSM-Playwright-scanner-pages-and-more response import fixed

Commit:
DSM-Playwright-scanner-pages-and-more able to enter multiple scan pairs

Commit:
DSM-Playwright-scanner-pages-and-more linter-fix

GiCharkviani committed:
Merge branch 'develop' into DSM-Playwright-scanner-pages-and-more

GiCharkviani committed:
Merge branch 'develop' into DSM-Playwright-scanner-pages-and-more

Commit:
DSM-Playwright-scanner-pages-and-more-remake fix

Commit:
DSM-Playwright-scanner-pages-and-more-remake removed duplicate identifier

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit first step

Commit:
Merge branch 'develop' into DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit moving forward

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit final scan covered

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit samples started

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit samples done

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit  tests implementation started

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit samples information assertion

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit tiny

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit tiny 2

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit todo

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit close to end

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fixes

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit removed only

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit added functional

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit changed small details

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linux images added

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit removed manual linux screenshots

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit kits without label for linux

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit temp skip

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit back to old one

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit rename to linux

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit added linux kit without label

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit added linux screenshot for kits sent table header

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit improved

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit kits received added

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit updated samples tab assertion

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit  throws differenterror on JSON parse

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit provided request object

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit finding the right participant

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit BSP_TOKEN added

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linux image added

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit BSPs

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit .sh file updated

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit skipping cohort tag tests

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit removed comment

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit updated comment

aweng98 commented on playwright-e2e/lib/component/dsm/filters/sections/customize-view.ts:
```diff
@@ -11,17 +11,27 @@ export class CustomizeView {
     const column = this.page.locator(columnXPath);
 
-    !(await this.isChecked(column)) && (await this.page.locator(columnXPath).click());
+    !deselect ? !(await this.isChecked(column)) && await this.page.locator(columnXPath).click() :
+      await this.isChecked(column) && await this.page.locator(columnXPath).click();
```
Since there are two calls each, I'd prefer to use if/else for clarity. What do you think?

GiCharkviani replied:
Otherwise, we will have nested if/else. Oki, I will do as you suggest.

aweng98 commented on playwright-e2e/lib/component/dsm/filters/sections/search/search.ts:
```diff
@@ -56,6 +59,11 @@ export class Search {
     textValue && !exactMatch && (await this.setExactMatch(columnName, true));
   }
 
+  public async radioBtn(columnName: string, { radioButtonValue, additionalFilters}: Partial<RadioButtonConfig>): Promise<void> {
```
nit: Change to `radiobutton` in order to keep name consistent with `widget/radiobutton` class.

aweng98 reviewed (commented):
This is awesome. Lots of changes and some good refactoring. 
I ran test locally. Screenshots comparisons has failed for me on mac. If you're using an external monitor to run test and generating screenshots, screenshot comparisons will fail for all others. Rerun test with `-u` flag without connected to an external monitor.

aweng98 commented on playwright-e2e/tests/dsm/kitUploadFlow/OS_PE-CGS-saliva-kit-upload-send-receive.spec.ts:
```diff
@@ -0,0 +1,179 @@
+  const kitType = KitTypeEnum.SALIVA;
+  const expectedKitTypes = [KitTypeEnum.SALIVA, KitTypeEnum.BLOOD];
+
+  test.beforeEach(async ({ page, request }) => {
```
I don't know why you created two `test.beforeEach`. Here and line 48.

GiCharkviani replied:
The first assigns and does initial preparation and the second calls methods on the initial (home) page.

aweng98 commented on playwright-e2e/lib/component/dsm/navigation/navigation.ts:
```diff
@@ -1,33 +1,56 @@
 
-  async selectFromStudy<T extends object>(studyNav: StudyNav): Promise<T> {
-    await this.selectFrom(MainMenu.STUDY, studyNav);
+  public async selectFromStudy<T extends object>(studyNav: StudyNavEnum): Promise<T> {
```
Cool generic function. Too bad none of DSM page extends a base dsm common page. something like `<T extends DSMPageBase>`

GiCharkviani replied:
Good idea. We might think. At this moment there is not any field or method that is the same across all DSM pages, except the page itself.

aweng98 commented on playwright-e2e/pages/dsm/kitsInfo-pages/kitsReceived-page/kitsReceivedPage.ts:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  public async waitForLoad(): Promise<void> {
+    await this.page.waitForLoadState('networkidle');
```
This is not an reliable way for synchronization. I'm good with keeping it for now because I don't have a better suggestion at this time.

aweng98 commented on playwright-e2e/pages/dsm/kitsInfo-pages/kitsWithoutLabel-page.ts:
```diff
@@ -0,0 +1,111 @@
+
+  private async deactivateKit(deactivateButton: Locator): Promise<void> {
+    await deactivateButton.click();
+    const deactivateReasonInput = await this.page.locator(this.deactivateReasonInputXPath);
```
Locator is not a async

GiCharkviani replied:
Ah, missed that. Thanks. Fixed!

aweng98 commented on playwright-e2e/pages/dsm/kitsInfo-pages/kitsWithoutLabel-page.ts:
```diff
@@ -0,0 +1,111 @@
+    await expect(await deactivateReasonInput,
+      'Kits Without Label page - Deactivate reason input field is not visible')
+      .toBeVisible();
+    await expect(await deactivateReasonButton,
```
Remove `await` in `await expect` calls.

GiCharkviani replied:
Done!

aweng98 commented on playwright-e2e/pages/dsm/participant-page/participant-page.ts:
```diff
@@ -0,0 +1,127 @@
+    return await this.tabs.clickTab<T>(tabName) as T;
+  }
+
+  private async readMainTextInfoFor(key: MainInfoEnum): Promise<string | null> {
+   return this.page.locator(this.getMainTextInfoXPath(key)).textContent();
+  }
```
This should be an non-async func.

GiCharkviani replied:
Good catch! Thanks. Done!

aweng98 commented on playwright-e2e/pages/dsm/participant-page/participant-page.ts:
```diff
@@ -0,0 +1,127 @@
+   return this.page.locator(this.getMainTextInfoXPath(key)).textContent();
+  }
+
+  private async readMainInputValueFor(key: MainInfoEnum): Promise<string | null> {
```
non-async func

GiCharkviani replied:
Done!

aweng98 commented on playwright-e2e/pages/dsm/participant-page/participant-page.ts:
```diff
@@ -0,0 +1,127 @@
+    return this.page.locator(this.getMainInputValueInfoXPath(key)).inputValue();
+  }
+
+  private async readMainCheckboxValueFor(key: MainInfoEnum): Promise<boolean> {
```
non-async

GiCharkviani replied:
Done!

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit after comments

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit updated snapshots

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linux screenshots updated

GiCharkviani commented:
@aweng98 Thanks for your reviews. I fixed all!
But, the tests still fail when I run them against test env - the cause is screenshot comparison. even though I did as you suggested (Unplugged the external monitor) and updated screenshots.

GiCharkviani committed:
Merge branch 'develop' into DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit fixed failing tests

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit skip cohort

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit updated tables headers assertions

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit linter fix

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit fix

aweng98 commented on playwright-e2e/utils/file-utils.ts:
```diff
@@ -8,6 +8,26 @@ export interface MailListCSV {
   lastName?: string;
 }
 
+export function crateTextFileSync(pathName: string, data: string) {
```
typo

GiCharkviani replied:
Thanks, fixed

aweng98 commented on playwright-e2e/tests/rgp/mailing-list-signup.spec.ts:
```diff
@@ -0,0 +1,68 @@
+import {expect} from '@playwright/test';
```
`rgp/mailing-list-signup.spec` have been renamed and moved. no longer exists.

GiCharkviani replied:
removed!

Commit:
DSM-PW-PEPPER-777--Kit-Upload_Send-and-Receive-a-Clinical-Saliva-Kit updates
